### DISCUSSION TITLE
Creating a prisma lambda layer with CDK to share across functions

### DISCUSSION THREAD

**Discussion by revmischa** [prisma/prisma#8329] [Q&A]
Continuing on the thread from here:

_Originally posted by @janpio in https://github.com/prisma/prisma/discussions/8245#discussioncomment-1012623_

> That I do not know much about honestly. Layers and CDK are quite advanced, so everyone seems to have very specific requirements that make it hard to generalize the solutions. If you get to one, please make sure to share it. Having a simple way to put your generated Client into a layer that is used across functions sounds like an awesome improvement in that regard.

I have been working on creating a reusable CDK lambda `LayerVersion` construct. The idea is that if you're building a CDK app with lambda functions that should use Prisma, they need to share the generated client and engine. Using a layer can reduce the bundle time and size for your functions because you can factor those out into the layer. The goal is to dynamically construct a lambda layer with the common dependencies and client.

This is where I'm at right now. We simply copy over [`prisma`, `.prisma`, `@prisma/client`] from the project's `node_modules` along with the `prisma` dir. Then inside the [layer docker bundling](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda.Code.html#static-fromwbrassetpath-options) we run `npx prisma generate` just to make sure we have the latest client. It all gets packaged as a layer that's added to every function.

<img width="1964" alt="Screen Shot 2021-07-17 at 2 33 56 PM" src="https://user-images.githubusercontent.com/245131/126035551-b8d88ef3-bf74-48fd-81b7-36c238936a98.png">

The output from the bundling looks like:
```
start layer
[
  'mkdir -p "/asset-output/nodejs/node_modules/prisma" && cp -rT "node_modules/prisma" "/asset-output/nodejs/node_modules/prisma"',
  'mkdir -p "/asset-output/nodejs/node_modules/.prisma" && cp -rT "node_modules/.prisma" "/asset-output/nodejs/node_modules/.prisma"',
  'mkdir -p "/asset-output/nodejs/node_modules/@prisma/client" && cp -rT "node_modules/@prisma/client" "/asset-output/nodejs/node_modules/@prisma/client"',
  'cp -r packages/repo/prisma /asset-output/nodejs/',
  'pushd /asset-output/nodejs',
  'HOME=/tmp npx prisma generate',
  'popd',
  'rm -f /asset-output/nodejs/node_modules/.prisma/client/*-engine-*',
  'rm -f /asset-output/nodejs/node_modules/prisma/client/*-engine-*',
  'rm -f /asset-output/nodejs/node_modules/prisma/*-engine-darwin',
  'rm -rf /asset-output/nodejs/node_modules/prisma/build'
]
Bundling asset freezy-mish/AppLayer/Code/Stage...
/asset-output/nodejs /asset-input
Prisma schema loaded from prisma/schema.prisma

✔ Generated Prisma Client (2.27.0) to ./node_modules/@prisma/client in 575ms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
.....
/asset-input
Bundling asset freezy-mish/MigrationScript/Code/Stage...

  packages/infra/cdk.out/bundling-temp-c42bcb6f2ee8700dcc0d89b156dabf6c1bda618ca67790630f275cbba4c8a52c/index.js  2.5mb ⚠️

⚡ Done in 180ms
Bundling asset freezy-mish/Gen/Func-blahblah-1/Code/Stage...

  packages/infra/cdk.out/bundling-temp-2b9daa95f3ad30dc12e29fd3c69f05042f0a4539a176730ff6641f006be74ade/index.js  2.8mb ⚠️

⚡ Done in 348ms
Bundling asset freezy-mish/Gen/Func-ContactAPI-2/Code/Stage...

  packages/infra/cdk.out/bundling-temp-84e58444f97414ac184d7262fccd660a66476e18a2a2c82c898cbe2881cc01e2/index.js  2.7mb ⚠️
```

The output bundle looks roughly good to me:
<img width="602" alt="Screen Shot 2021-07-17 at 2 40 44 PM" src="https://user-images.githubusercontent.com/245131/126035700-cfe1d7eb-ed66-48c5-9592-6ec8b459a7d9.png">

And I think it should all work. But when I invoke my lambda I get:
```ts
        "Error: @prisma/client did not initialize yet. Please run \"prisma generate\" and try to import it again.",
        "In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues",
        "    at new PrismaClient (/opt/nodejs/node_modules/.prisma/client/index.js:3:11)",
        "    at /var/task/index.js:95:5291",
        "    at /var/task/index.js:1:861",
        "    at /var/task/index.js:95:80530",
        "    at /var/task/index.js:1:861",
        "    at Object.<anonymous> (/var/task/index.js:95:82179)",
        "    at Module._compile (internal/modules/cjs/loader.js:1068:30)",
        "    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)",
        "    at Module.load (internal/modules/cjs/loader.js:933:32)",
        "    at Function.Module._load (internal/modules/cjs/loader.js:768:16)"
```

All of the magic bundling and differences between the three packages is a bit opaque to me. I've been struggling a lot all this week trying to figure out how to work around the weird prisma tricks that happen in node_modules when packaging up lambda functions. Please let me know what I'm doing wrong.

**Comment by revmischa** [comment]
Been working on this for a few days and still not able to get this to work.
Currently I'm trying this to package my layer:
```ts
[
  'cp -r packages/repo/prisma /asset-output/nodejs/',
  'pushd /asset-output/nodejs',
  'HOME=/tmp npx prisma generate',
  'popd',
  'rm -f /asset-output/nodejs/node_modules/.prisma/client/*-engine-*',
  'rm -f /asset-output/nodejs/node_modules/prisma/client/*-engine-*',
  'rm -f /asset-output/nodejs/node_modules/@prisma/engines/prisma-fmt-*',
  'rm -f /asset-output/nodejs/node_modules/@prisma/engines/introspection-engine-*',
  'rm -f /asset-output/nodejs/node_modules/@prisma/engines/migration-engine-*',
  'rm -f /asset-output/nodejs/node_modules/prisma/*-engine-darwin',
  'rm -rf /asset-output/nodejs/node_modules/prisma/build',
  'rm -rf /asset-output/nodejs/node_modules/.bin'
]
```

This generates a layer that looks right to me (though I have no clue what is "right" or why there are three prisma directories and the magic that goes with them).

It contains `nodejs/node_modules/{prisma,@prisma,.prisma}`. My generated client is in `.prisma/client/index.js` - I've confirmed it has my models.

This layer is attached to my functions. Since it has the `.prisma` and `@prisma/client` modules I would expect to not get this error saying I need to run `prisma generate`.

Attached is my layer zip.

[freezy-mish-app-7d004d6e-95cd-4c56-ba7d-2a469716937d.zip](https://github.com/prisma/prisma/files/6835874/freezy-mish-app-7d004d6e-95cd-4c56-ba7d-2a469716937d.zip)

**Comment by hhimanshu** [comment]
Hey, do you have a github repo that I can see to learn more. I have a similar use case. thank you!

**Reply by revmischa** [reply]
Something like this https://github.com/jetbridge/sst-prisma/blob/master/stacks/resources/prismaLayer.ts
Need to update for 4.8.0

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
